File: java-impl/src/main/java/hackathonlib/Network.java
package hackathonlib;

import java.net.ServerSocket;
import java.net.Socket;
import java.util.LinkedList;
import java.util.List;

/**
 * Models a network for servers. Accepts clients on ports.
 * @author Michael Fürst
 * @version 1.0
 * @since 23.11.2014
 */
public class Network implements Runnable{
    private boolean running = false;
    private ServerSocket server;

    private int port;
    private IGameManager gameManager;
    private int receiveTimeout;

    private Thread th = null;
    private List<Connection> threads = new LinkedList<>();

    /**
     * Start a server service on a given port using the game manager.
     * @param port The port.
     * @param gameManager The gameManager.
     * @param receiveTimeout The timeout how often clients must be active before kicked in seconds.
     */
    public void startListen(int port, IGameManager gameManager, int receiveTimeout) {
        if (th == null) {
            this.gameManager = gameManager;
            this.receiveTimeout = receiveTimeout;
            th = new Thread(this);
            th.start();
        }
    }

    @Override
    public void run() {
        running = true;

        try {
            server = new ServerSocket(port);
            server.setSoTimeout(10);
        } catch (Exception e) {
            e.printStackTrace();
        }

        while (running) {
            try {
                Socket c = server.accept();
                if (c == null) continue;
                threads.add(new Connection(c, gameManager, receiveTimeout));
            } catch (Exception e) {
                // Do nothing...
            }
        }
        try {
            server.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        server = null;
        threads.forEach(t -> t.quit());
        th = null;
    }

    /**
     * Stop listening for clients and disconnect all connected clients.
     */
    public void endListen() {
        running = false;
    }
}

File: csharp-impl/HackathonLib/Network.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace Hackathonlib
{
    /// <summary>
    /// A class that encapsulates a connection listener.
    /// </summary>
    public class Network
    {
        private int port;
        private IGameManager gameManager;
        private int receiveTimeout;

        private TcpListener listener;
        private List<Connection> threads = new List<Connection>();
        private bool running = true;
        private Thread th;

        /// <summary>
        /// Connect to the server. Automatically retries when not successfull.
        /// </summary>
        /// <param name="hostname"></param>
        /// <param name="port"></param>
        /// <returns></returns>
        public static TcpClient Connect(string hostname, int port)
        {
            TcpClient client = null;
            while (client == null)
            {
                try
                {
                    client = new TcpClient(hostname, port);
                }
                catch (Exception)
                {
                }
            }
            return client;
        }

        /// <summary>
        /// Start to listen for incoming connections.
        /// </summary>
        /// <param name="port">Listen on this port.</param>
        public void StartListen(int port, IGameManager gameManager, int receiveTimeout)
        {
            this.port = port;
            this.gameManager = gameManager;
            this.receiveTimeout = receiveTimeout;

            th = new Thread(new ThreadStart(Run));
            th.Start();
        }

        private void Run()
        {
            try
            {
                listener = new TcpListener(port);
                listener.Start();
            }
            catch (Exception)
            {
                Console.Error.WriteLine("Cannot open port...");
            }

            while (running)
            {
                try
                {
                    // Wartet auf eingehenden Verbindungswunsch
                    TcpClient c = listener.AcceptTcpClient();
                    if (c == null) continue;
                    // Initialisiert und startet einen Server-Thread
                    // und fügt ihn zur Liste der Server-Threads hinzu
                    threads.Add(new Connection(c, gameManager, receiveTimeout, false));
                }
                catch (Exception)
                {
                    // Do nothing..
                }
            }

            listener.Stop();
            listener = null;

            // Alle Server-Threads stoppen
            foreach (var t in threads)
            {
                t.Quit();
            }
        }

        /// <summary>
        /// Stop listening on the network.
        /// </summary>
        public void EndListen()
        {
            running = false;
            if (listener != null)
            {
                var conn = new TcpClient("localhost", port);
                conn.Close();
            }
        }
    }
}
